Wrappers/Python/cil/processors/PaganinProcessor.py

```diff
@@ -204,101 +205,78 @@ def check_input(self, data):
         
         if data.dtype!=np.float32:
             raise TypeError('Processor only support dtype=float32')
+        
+        cil_order = tuple(AcquisitionDimension.get_order_for_engine('cil',data.geometry))
+        if data.dimension_labels != cil_order:
+            raise ValueError("This processor requires CIL data order, consider using `data.reorder('cil')`")
     
         return True
 
     def process(self, out=None):
 
-        data  = self.get_input()
-        cil_order = tuple(AcquisitionDimension.get_order_for_engine('cil',data.geometry))
-        if data.dimension_labels != cil_order:
-            log.warning(msg="This processor will work most efficiently using\
-                        \nCIL data order, consider using `data.reorder('cil')`")
-
+        data  = self.get_input()            
         # set the geometry parameters to use from data.geometry unless the
         # geometry is overridden with an override_geometry
         self._set_geometry(data.geometry, self.override_geometry)
 
         if out is None:
             out = data.geometry.allocate(None)
 
-        # make slice indices to get the projection
-        slice_proj = [slice(None)]*len(data.shape)
-        angle_axis = data.get_dimension_axis('angle')
-        slice_proj[angle_axis] = 0
-
-        if data.geometry.channels>1:
-            channel_axis = data.get_dimension_axis('channel')
-            slice_proj[channel_axis] = 0
-        else:
-            channel_axis = None
-
-        data_proj = data.as_array()[tuple(slice_proj)]
-
         # create an empty axis if the data is 2D
-        if len(data.shape) == 2:
-            data.array = np.expand_dims(data.array, len(data.shape))
-            slice_proj.append(slice(None))
-            data_proj = data.as_array()[tuple(slice_proj)]
-
-        elif len(data_proj.shape) == 2:
-            pass
-        else:
+        if data.geometry.dimension == '2D':
+            data.array = np.expand_dims(data.array, axis=-1)
+            if len(out.shape) == 2:
+                out.array = np.expand_dims(out.array, axis=-1)
+        # expand dimensions for single channels
+        if data.geometry.channels == 1:
+            data.array = np.expand_dims(data.array, axis=0)
+            if out.array.ndim < data.array.ndim:
+                out.array = np.expand_dims(out.array, axis=0)
+        # expand dimensions for single angles
+        if len(data.geometry.angles) == 1:
+            data.array = np.expand_dims(data.array, axis=1) 
+            if out.array.ndim < data.array.ndim:
+                out.array = np.expand_dims(out.array, axis=1)
```

Comment:
here can you calculate the shape you need and then just use reshape to set it to that, I think it would be clearer to see what's happening and what shape you're tying to achieve.

Reply:
Think I've updated this now